src/lib/increment.js

```diff
@@ -0,0 +1,38 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+module.exports = function incrementNumeric(x, delta) {
+    if(!delta) return x;
+
+    // Note 1:
+    // 0.3 != 0.1 + 0.2 == 0.30000000000000004
+    // but 0.3 == (10 * 0.1 + 10 * 0.2) / 10
+    // Attempt to use integer steps to increment
+    var scale = 1 / Math.abs(delta);
+    if(scale < 1) scale = 1;
+    var newX = (
+        scale * x +
+        scale * delta
+    ) / scale;
+
+    // Note 2:
+    // now we may also consider rounding to cover few more edge cases
+    // e.g. 0.3 * 3 = 0.8999999999999999
+    var lenDt = ('' + delta).length;
+    var lenX0 = ('' + x).length;
+    var lenX1 = ('' + newX).length;
+
+    if(lenX1 >= lenX0 + lenDt) { // likely a rounding error!
+        newX = +parseFloat(newX).toPrecision(12);
+    }
```

Comment:
This is quite a bit of work to do every time - any way we can reduce it? Maybe start with `lenX1` and only do the rest if that's `>=6` or something?

Also I'm uneasy about what this will do with very large or very small numbers, particularly when crossing the points where `toPrecision` returns exponential notation (e-7 or less, and e+21 or more). I guess the only really problematic case I can see is a very tight range right around -1e+21:
```js
x0 = -1e+21 // length=6
delta = 1000000 // length 7
x1 = x0 + delta // length 22 > 13 (-999999999999999000000)
+parseFloat(x1).toPrecision(12) // rounds to -1e+21 -> you get stuck
```

Maybe we just skip this check when `x` or `newX` is sufficiently large?

Also a little worried about if we misround a tick that's intended to be zero - but I don't see a way this will cause problems there, just ways it'll fail to fix a problem we already have, which is fine 😅

Reply:
> This is quite a bit of work to do every time - any way we can reduce it? Maybe start with `lenX1` and only do the rest if that's `>=6` or something?

Good call. Done in https://github.com/plotly/plotly.js/pull/5114/commits/789e7e31f523ac32b244514a7ba3ca992d394da5. Now that step only applies when `lenX1 > 16`.

Reply:
> Also I'm uneasy about what this will do with very large or very small numbers, particularly when crossing the points where `toPrecision` returns exponential notation (e-7 or less, and e+21 or more). I guess the only really problematic case I can see is a very tight range right around -1e+21:
> 
> ```js
> x0 = -1e+21 // length=6
> delta = 1000000 // length 7
> x1 = x0 + delta // length 22 > 13 (-999999999999999000000)
> +parseFloat(x1).toPrecision(12) // rounds to -1e+21 -> you get stuck
> ```
> 
> Maybe we just skip this check when `x` or `newX` is sufficiently large?

Good call. 
Addressed in https://github.com/plotly/plotly.js/pull/5114/commits/ebb63c815e5edff65ed81d91b2deef36b296235b & https://github.com/plotly/plotly.js/pull/5114/commits/671b57884cef954ad5a5cb868ad64a78212b13f2  and tested in https://github.com/plotly/plotly.js/pull/5114/commits/27f9774d32fbc4717091f3725106b531fb794161.